File: turgles/random_walk.c
#include <math.h>
#include <stdlib.h>
#include <time.h>

#define TURTLE_DATA_SIZE 12

void random_walk(float *x, float magnitude, float half_size, float lambda, float degrees)

{
    float *y = x + 1;
    float *angle = x + 4;
    float *ct = x + 6;
    float *st = x + 7;
    if (fabs(*x) > half_size || fabs(*y) > half_size)
    {
        *angle = fmod(*angle + 180.0, 360.0);
    }
    float r = rand() / (float)(RAND_MAX);
    float rd = (r * 2 * degrees) - degrees;
    *angle = fmod(*angle + rd, 360.0);
    float theta = *angle / 180.0 * M_PI;
    *ct = cos(theta);
    *st = sin(theta);
    *x += magnitude * *ct;
    *y += magnitude * *st;
}

void random_walk_all(
        float *turtles, 
        int num_turtles, 
        float magnitude, 
        float half_size,
        float lambda,
        float degrees
)
{
    int i;
    for (i = 0; i < num_turtles; ++i)
    {
        random_walk(turtles + (i * TURTLE_DATA_SIZE), magnitude, half_size, lambda, degrees);
    }
}

File: turgles/random_walk.py
from __future__ import division, print_function, absolute_import

from random import random
from math import radians, sin, cos

from turgles.config import speed, degrees, world_width, world_height
from turgles.memory import ffi, TURTLE_MODEL_DATA_SIZE

half_w = world_width // 2
half_h = world_height // 2

ffi.cdef(
    "void random_walk_all(float*, int, float, float, float, float);"
)
try:
    fast = ffi.dlopen('./libfast.so')
except:
    fast = None


def slow_update(dt, buffers):
    magnitude = speed * dt
    for b in buffers:
        walk(b.model.data, b.count, magnitude, half_w, degrees)

if fast:
    def fast_update(dt, buffers):
        magnitude = speed * dt
        for b in buffers:
            fast.random_walk_all(
                b.model.data, b.count, magnitude, half_w, half_h, degrees)
    update = fast_update
else:
    update = slow_update


def walk(data, size, magnitude, scale, degress):
    for x in range(0, size * TURTLE_MODEL_DATA_SIZE, TURTLE_MODEL_DATA_SIZE):
        y = x + 1
        a = x + 4
        angle = data[a]
        absx, absy = abs(data[x]), abs(data[y])
        if absx > half_h or absy > half_w:
            angle = (angle + 180) % 360
        angle += (random() * 2 * degrees) - degrees
        theta = radians(angle)
        ct = cos(theta)
        st = sin(theta)
        dx = magnitude * ct
        dy = magnitude * st
        data[x] += dx
        data[y] += dy
        data[a] = angle
        data[x + 6] = ct
        data[x + 7] = st
